lib/vfscore/automount.c

```diff
@@ -0,0 +1,309 @@
+
+#ifndef CONFIG_LIBVFSCORE_ROOTFLAGS
+static __u64 rootflags;
+#else
+static __u64 rootflags      = (__u64) CONFIG_LIBVFSCORE_ROOTFLAGS;
+#endif
+
+#endif /* CONFIG_LIBVFSCORE_AUTOMOUNT_ROOTFS */
+
+#ifdef CONFIG_LIBVFSCORE_FSTAB
+
+#define LIBVFSCORE_FSTAB_VOLUME_ARGS_SEP			':'
+
+static char *vfscore_fstab[CONFIG_LIBVFSCORE_FSTAB_SIZE];
+
+UK_LIBPARAM_PARAM_ARR_ALIAS(fstab, &vfscore_fstab, charp,
+			    CONFIG_LIBVFSCORE_FSTAB_SIZE,
+			    "VFSCore Filesystem Table");
+
+/**
+ * Expected command-line argument format:
+ *	vfs.fstab=[
+ *		"<src_dev>:<mntpoint>:<fsdriver>[:<flags>:<opts>]"
+ *		"<src_dev>:<mntpoint>:<fsdriver>[:<flags>:<opts>]"
+ *		...
+ *	]
+ * These list elements are expected to be separated by whitespaces.
+ * Mount options and flags are optional.
+ */
+static char *vfscore_fstab_next_volume_arg(char **pos)
+{
+	char *ipos;
+
+	UK_ASSERT(pos);
+
+	if (!*pos)
+		return NULL;
+
+	ipos = *pos;
+	*pos = strchr(ipos, LIBVFSCORE_FSTAB_VOLUME_ARGS_SEP);
+
+	if (*pos && **pos != '\0') {
+		**pos = '\0';
+		(*pos)++;
+	} else {
+		*pos = NULL;
+	}
+
+	return ipos;
+}
+
+static void vfscore_fstab_fetch_volume_args(char *v, struct vfscore_volume *vv)
+{
+	const char *flags;
+	char *pos;
+
+	UK_ASSERT(v);
+	UK_ASSERT(vv);
+
+	pos = v;
+
+	/* sdev, path and drv are mandatory */
+	vv->sdev = vfscore_fstab_next_volume_arg(&pos);
+	UK_ASSERT(vv->sdev);
+
+	vv->path = vfscore_fstab_next_volume_arg(&pos);
+	UK_ASSERT(vv->path);
+
+	vv->drv = vfscore_fstab_next_volume_arg(&pos);
+	UK_ASSERT(vv->drv);
+
+	if (!pos)
+		return;
+
+	/* Allow empty flags: "<src_dev>:<mntpoint>:<fsdriver>::<opts>" */
+	flags = vfscore_fstab_next_volume_arg(&pos);
+	if (flags && flags[0] != '\0')
+		vv->flags = strtol(flags, NULL, 0);
+	else
+		vv->flags = 0;
+
+	vv->opts = vfscore_fstab_next_volume_arg(&pos);
+	if (vv->opts && vv->opts[0] == '\0')
+		vv->opts = NULL;
+}
+#endif /* CONFIG_LIBVFSCORE_FSTAB */
+
+#ifdef CONFIG_LIBUKCPIO
+static int vfscore_mount_initrd_volume(struct vfscore_volume *vv)
+{
+	struct ukplat_memregion_desc *initrd;
+	int rc;
+
+	UK_ASSERT(vv);
+
+	/* TODO: Support multiple Initial RAM Disks */
+	rc = ukplat_memregion_find_initrd0(&initrd);
+	if (unlikely(rc < 0)) {
+		uk_pr_crit("Could not find an initrd!\n");
+
+		return -1;
+	}
+
+	uk_pr_info("Mount ramfs to /...\n");
+
+	rc = mount(vv->sdev, vv->path, "ramfs", vv->flags, vv->opts);
+	if (unlikely(rc)) {
+		uk_pr_crit("Failed to mount ramfs to /: %d\n", errno);
+
+		return -1;
+	}
+
+	uk_pr_info("Extracting initrd @ %p (%"__PRIsz" bytes) to %s...\n",
+		   (void *)initrd->vbase, initrd->len, vv->path);
+
+	rc = ukcpio_extract(vv->path, (void *)initrd->vbase, initrd->len);
+	if (unlikely(rc)) {
+		uk_pr_crit("Failed to extract cpio archive to %s: %d\n",
+			   vv->path, rc);
+
+		return -1;
+	}
+
+	return 0;
+}
+
+#else /* CONFIG_LIBUKCPIO */
+static int vfscore_mount_initrd_volume(struct vfscore_volume *vv __unused)
+{
+	uk_pr_warn("Cannot extract InitRD. Enable CONFIG_LIBRAMFS.\n");
```

Comment:
Oops should have been `CONFIG_LIBUKCPIO`

Reply:
Probably we should just fall back to a default message, something like saying: `Unknown filesystem: initrd`. I guess we have something like this already built-in with `mount()`. So, I think a separate handler for "initrd-support-is-not-present" is not needed.